### DISCUSSION TITLE
Validate a PURL

### DISCUSSION THREAD

**Discussion by visagansanthanam-unisys** [package-url/purl-spec#296] [Q&A]
Hi,

We are trying to generate a purl for the components. Is there a way to find and validate a pURL for an existing component? how to make sure the pURL is valid?

**Comment by stevespringett** [comment]
I think all official purl implementations have the ability to validate a purl string to ensure it is compliant with the standard. I'm not aware of any tool that can verify that a specific component artifact has a specific purl. Perhaps some commercial tooling does this. I'm not entirely sure.

**Comment by pombredanne** [comment]
@visagansanthanam-unisys 
As @stevespringett mentioned, you can validate the syntax in most libraries. 

Beyond this, some FOSS stuffs that may be some help:
 
- The https://github.com/nexB/purldb/ app is to builds a large DB of code keyed by PURL, and there is a first demo API at https://public.purldb.io/api/
- The https://github.com/nexB/purldb/blob/main/purldb-toolkit/ is a new command line util and library to perform various queries and validations on a purl, including both syntax and existence validation  
- For the inverse, https://github.com/package-url/packageurl-python/blob/main/src/packageurl/contrib/url2purl.py is trying to build a PURL from a URL. 
- https://github.com/nexB/fetchcode/ takes a PURL as an input and fetches the code behind.

I hope this helps!

**Comment by pombredanne** [comment]
I am moving this to a discussion BTW.

**Comment by andrewpollock** [comment]
FYI the CVE Program has an interest in being able to validate purls as part of inclusion of them in the CVE Schema.

See https://github.com/CVEProject/cve-schema/pull/407

**Reply by pombredanne** [reply]
@andrewpollock Thanks!
I am working on some extensive tooling and data for PURL validation!

See these many issues to track the work specifically on PURL validation:
- https://github.com/orgs/aboutcode-org/projects/52/views/22?filterQuery=projectid%3Apurl-next+issueid%3Apurl-next-m3*%2Cpurl-next-m5*
- https://github.com/orgs/aboutcode-org/projects/52/views/10
- https://github.com/orgs/aboutcode-org/projects/52/views/21
- https://github.com/orgs/aboutcode-org/projects/54

In particular:
- validate PURL overall https://github.com/aboutcode-org/purl-spec/issues/43
- create a data-driven offline validator in Rust/Go/Python: 
  - https://github.com/aboutcode-org/purl-validator/issues/18
  - https://github.com/aboutcode-org/purl-validator/issues/20
- same but online: https://github.com/aboutcode-org/purldb/issues/643
- adding extended validation in Python: https://github.com/aboutcode-org/purl-spec/issues/44
- validate SBOMs https://github.com/aboutcode-org/scancode.io/issues/1597 
- help validate accuracy of scanner/tools, including for binaries:
  - https://github.com/aboutcode-org/purl-spec/issues/48
  - https://github.com/aboutcode-org/purl-spec/issues/66
  - https://github.com/aboutcode-org/scancode.io/issues/1770
  - https://github.com/aboutcode-org/purl-spec/issues/61

Also, help major FOSS foundations and registries adopt/use the tools:
- https://github.com/aboutcode-org/purl-spec/issues/56

(separately crates.io now displays PURL on each crate's page)

@andrewpollock so as you can se there is a lot of things planned. A shit load of work to do! Help is mucho wanted :angel: 

@alilleybrinker FYI

**Reply by alilleybrinker** [reply]
Thanks @pombredanne, this is great!

For more context on the CVE QWG's thinking: right now we're planning to, at minimum, require submitted PURLs to parse with a first-party PURL library, and require that they not include a version (since the CVE Record Format already has a defined way to specify versions).

The opportunity for stronger validation, or to benefit from improvements to implementations based on the spec-compliance testing approach some of the linked issues describe, would be great!

Looking forward to how this develops!
